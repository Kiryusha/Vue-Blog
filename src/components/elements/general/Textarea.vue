<template lang="pug">
  .textarea(
    :class="['input', {'_valid': valid && !error, '_error': error}, cssClass]"
  )
    textarea(
      ref="textarea"
      :rows="rows"
      :placeholder="placeholder",
      :name="name",
      :disabled="disabled",
      :maxlength="maxlength",
      v-model="innerValue",
      class="textarea__control",
      @input="inputHandler",
      @blur="$emit('blur')",
      @focus="$emit('focus')"
    )
    .textarea__error(v-if="error") {{ error }}
</template>

<script>
export default {
  props: {
    rows: {
      type: Number,
      default: 1,
    },
    maxlength: {
      type: Number,
      default: 200,
    },
    placeholder: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
    error: {
      type: String,
      default: '',
    },
    value: {
      type: String,
      default: '',
    },
    cssClass: {
      type: String,
      default: '',
    },
    valid: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      innerValue: this.value,
    };
  },
  watch: {
    value(value) {
      this.innerValue = value;
    },
  },
  methods: {
    inputHandler() {
      const el = this.$refs.textarea;

      el.style.cssText = 'height:auto;';
      el.style.cssText = '-moz-box-sizing:content-box';
      el.style.cssText = `min-height:${el.scrollHeight}px`;

      this.$emit('input', this.innerValue);
    },
  },
};
</script>

<style lang="stylus">
  .textarea
    box-sizing border-box
    -webkit-tap-highlight-color rgba(0, 0, 0, 0)

    &__control
      font-size 18px
      line-height 1.5
      border 1px solid rgba(#000, .15)
      border-radius 4px
      padding 10px 15px
      -webkit-appearance none
      display block
      width 100%
      font-family inherit
      outline none
      resize none
      overflow hidden
      word-wrap break-word

    &._error &__control
      border-color red

    &__error
      font-size 16px
      color red
      padding-top 10px

    +phone()

      &__control
        font-size 16px
        padding 10px 10px

</style>
