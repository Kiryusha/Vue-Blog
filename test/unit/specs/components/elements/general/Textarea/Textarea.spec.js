// Textarea.vue

import { shallowMount } from '@vue/test-utils';
import Textarea from 'Components/elements/general/Textarea.vue';

describe('Textarea.vue', () => {

  describe('component shallow rendering', () => {
    it('should match a snapshot', () => {
      const wrapper = shallowMount(Textarea);

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should match a snapshot - with error', () => {
      const wrapper = shallowMount(Textarea);

      wrapper.setProps({
        error: 'test error',
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('lifecycle-driven methods calls', () => {
    it('should calln inputHandler() on an update hook', () => {
      const inputHandler = jest.fn();
      const wrapper = shallowMount(Textarea, { methods: {
        inputHandler,
      }});

      wrapper.setData({
        innerValue: 'test',
      })

      expect(inputHandler).toHaveBeenCalledTimes(1);
      });
  });

  describe('custom actions', () => {
    it('should set innerValue on value prop change', () => {
      const wrapper = shallowMount(Textarea);

      wrapper.setProps({
        value: 'test',
      });

      expect(wrapper.vm.innerValue).toBe('test');
    });

    it('should emit parent input event on local input event', () => {
      const wrapper = shallowMount(Textarea);

      wrapper.find('.control').trigger('input');

      expect(wrapper.emitted().input).toBeTruthy();
    });
  });
});
