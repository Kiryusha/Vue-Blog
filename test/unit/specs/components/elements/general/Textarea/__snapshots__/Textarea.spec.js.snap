// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea.vue component shallow rendering should match a snapshot - with error 1`] = `
<div class="textarea input _error">
  <textarea rows="1" placeholder="" name="" maxlength="200" class="control" style="min-height: 0px;"></textarea>
  <div class="error">test error</div>
</div>
`;

exports[`Textarea.vue component shallow rendering should match a snapshot 1`] = `
<div class="textarea input">
  <textarea rows="1" placeholder="" name="" maxlength="200" class="control"></textarea>
  <!---->
</div>
`;
